lib/record_store/provider/ns1.rb

```diff
@@ -0,0 +1,222 @@
+        client.zones.map do |zone|
+          zone["zone"]
+        end
+      end
+
+      private
+
+      # Fetches simplified records for the provided zone
+      def records_for_zone(zone)
+        client.zone(zone)["records"]
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        new_answers = [{ answer: build_api_answer_from_record(record) }]
+
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type
+        )
+
+        if existing_record.nil?
+          client.create_record(
+            zone: zone,
+            fqdn: record_fqdn,
+            type: record.type,
+            params: { answers: new_answers, ttl: record.ttl }
+          )
+          return
+        end
+
+        existing_answers = existing_record["answers"].map { |answer| symbolize_keys(answer) }
+        client.modify_record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type,
+          params: { answers: existing_answers + new_answers, ttl: record.ttl }
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type
+        )
+        return if existing_record.nil?
+
+        pruned_answers = existing_record["answers"]
+          .map { |answer| symbolize_keys(answer) }
+          .reject { |answer| answer[:answer] == build_api_answer_from_record(record) }
+
+        if pruned_answers.empty?
+          client.delete_record(
+            zone: zone,
+            fqdn: record_fqdn,
+            type: record.type
+          )
+          return
+        end
+
+        client.modify_record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type,
+          params: { answers: pruned_answers }
+        )
+      end
+
+      # Updates an existing record in the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # id - provider specific ID of record to update
+      # record - a kind of `Record` which the record with `id` should be updated to
+      def update(id, record, zone)
+        record_fqdn = record.fqdn.sub(/.$/, '')
+
+        existing_record = client.record(
+          zone: zone,
+          fqdn: record_fqdn,
+          type: record.type,
+          must_exist: true,
+        )
+
+        # Identify the answer in this record with the matching ID, and update it
+        updated = false
+        existing_record["answers"].each do |answer|
+          if answer["id"] == id
```

Comment:
To gain an indentation level, you can revert the condition:

```suggestion
          next if answer["id"] != id
```

I would also add a `break` at the end of the loop to avoid checking all of the other records, as we only want the one matching the id.

Another possibility (probably even more ruby-like) would be:

```ruby
existing_record['answers'].select { |answer| answer['id'] == id }.each do |answer|
  updated = true
  answer['answer'] = build_api_answer_from_record(record)
end
```